### PR TITLE
Deprecate DataModule properties: train_transforms, val_transforms, test_transforms, dims, and size

### LINKED ISSUE #8728: Deprecate {train/val/test}_transforms, dims, and size from the DataModule
## 🚀 Feature

Deprecate these properties off the LightningDataModule interface:
https://github.com/PyTorchLightning/pytorch-lightning/blob/963c26764682fa4cf64c93c5a7572ae0040e9c32/pytorch_lightning/core/datamodule.py#L94-L147

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740#issuecomment-876780318

`train_transforms`, `val_transforms`, `test_transforms`, `dims`, and `size` are entirely optional to use, yet they’re on the DataModule interface. The Trainer does not rely on these, nor is the user forced to implement them. In reality, these are internal implementation details of individual datamodules. As a result, we ought to deprecate these off the DataModule interface.

Of course, users retain the ability to implement these properties in their LightningModules if they find these abstractions helpful.

### Pitch

We can follow a similar approach as what was done for https://github.com/PyTorchLightning/pytorch-lightning/issues/7301
- Mark the properties as deprecated in v1.5. Warn if users are passing non-None values to the DataModule constructor and if they read/set their corresponding properties
- Remove the properties in v1.7

### Alternatives
Keep as is

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by Tshimanga**
Deprecate DataModule properties: train_transforms, val_transforms, test_transforms, dums, and size

## What does this PR do?

added deprecation warnings to Datamodule methods: train_transforms, val_transforms, test_transforms, dums, and size.

as per the approach outlined in the ticket based on #7301 where the deprecation warning goes out next release and then is removed two releases later. I did not find any tests that need to be moved.

Fixes #8728 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -81,7 +81,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `LightningModule.model_size` ([#8343](https://github.com/PyTorchLightning/pytorch-lightning/pull/8343))
 
 
--
+- Deprecated `DataModule` properties: `train_transforms`, `val_transforms`, `test_transforms`, `size`, `dims`
```
Comment (ananthsub) [inline CHANGELOG.md:84]:
```suggestion
- Deprecated `DataModule` properties: `train_transforms`, `val_transforms`, `test_transforms`, `size`, `dims` ([#8851](https://github.com/PyTorchLightning/pytorch-lightning/pull/8851))
```

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -70,6 +70,18 @@ def teardown(self):
...
+                "DataModule property `train_transforms` was deprecated in v1.5 and will be removed in v1.7."
+            )
+        if val_transforms is not None:
+            rank_zero_deprecation(
+                "DataModule property `val_transforms` was deprecated in v1.5 and will be removed in v1.7."
+            )
+        if test_transforms is not None:
+            rank_zero_deprecation(
+                "DataModule property `test_transforms` was deprecated in v1.5 and will be removed in v1.7."
+            )
```
Comment (ananthsub) [inline pytorch_lightning/core/datamodule.py:84]: please add a warning for `dims` too

File: tests/deprecated_api/test_remove_1-7.py [hunk 3]
```diff
@@ -46,3 +47,25 @@ def test_v1_7_0_deprecated_model_size():
         match="LightningModule.model_size` property was deprecated in v1.5 and will be removed in v1.7"
     ):
         _ = model.model_size
+
+
+def test_v1_7_0_datamodule_transform_properties(tmpdir):
+    dm = MNISTDataModule()
```
Comment (ananthsub) [inline tests/deprecated_api/test_remove_1-7.py:53]: add a test with a dummy datamodule which passes the transforms & dims constructor arguments to the base DataModule, and confirm that the deprecation warning appears for those as well

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (Tshimanga) [reply pytorch_lightning/core/datamodule.py:84]: oh dear.. thanks for catching 👍🏽

File: tests/deprecated_api/test_remove_1-7.py [hunk 4]
```diff
@@ -46,3 +48,33 @@ def test_v1_7_0_deprecated_model_size():
...
+        dm.val_transforms = "b"
+    with pytest.deprecated_call(match=r"DataModule property `test_transforms` was deprecated in v1.5"):
+        dm.test_transforms = "c"
+    with pytest.deprecated_call(match=r"DataModule property `train_transforms` was deprecated in v1.5"):
+        _ = LightningDataModule(train_transforms="a")
+    with pytest.deprecated_call(match=r"DataModule property `val_transforms` was deprecated in v1.5"):
+        _ = LightningDataModule(val_transforms="b")
+    with pytest.deprecated_call(match=r"DataModule property `test_transforms` was deprecated in v1.5"):
+        _ = LightningDataModule(test_transforms="c")
+
```
Comment (ananthsub) [inline tests/deprecated_api/test_remove_1-7.py:67]: Include dims here too

**Review by ananthsub — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-7.py (hunk 4 above)
Comment (Tshimanga) [reply tests/deprecated_api/test_remove_1-7.py:67]: @ananthsub how come? test the non-None `dims` construction in the `test_v1_7_0_datamodule_dims_property` method. is there a reason to test them all together?

File: tests/deprecated_api/test_remove_1-7.py [hunk 5]
```diff
@@ -46,3 +48,33 @@ def test_v1_7_0_deprecated_model_size():
...
+    with pytest.deprecated_call(match=r"DataModule property `size` was deprecated in v1.5"):
+        dm.size()
+
+
+def test_v1_7_0_datamodule_dims_property(tmpdir):
+    dm = MNISTDataModule()
+    with pytest.deprecated_call(match=r"DataModule property `dims` was deprecated in v1.5"):
+        _ = dm.dims
+    with pytest.deprecated_call(match=r"DataModule property `dims` was deprecated in v1.5"):
+        _ = LightningDataModule(dims=(1, 1, 1))
```
Comment (Tshimanga) [inline tests/deprecated_api/test_remove_1-7.py:80]: here @ananthsub

**Comment by carmocca** [comment]
These are some occurrences found by grepping the project which should be addressed/updated/removed

```python
tests/helpers/boring_model.py:163:            self.dims = self.random_train[0].shape
tests/helpers/boring_model.py:170:            self.dims = getattr(self, "dims", self.random_test[0].shape)
tests/helpers/boring_model.py:174:            self.dims = getattr(self, "dims", self.random_predict[0].shape)
tests/helpers/datamodules.py:43:        # self.dims is returned when you call dm.size()
tests/helpers/datamodules.py:44:        # Setting default dims here because we know them.
tests/helpers/datamodules.py:46:        self.dims = (1, 28, 28)
docs/source/extensions/datamodules.rst:130:            # self.dims is returned when you call dm.size()
docs/source/extensions/datamodules.rst:131:            # Setting default dims here because we know them.
docs/source/extensions/datamodules.rst:133:            self.dims = (1, 28, 28)
docs/source/extensions/datamodules.rst:148:                # self.dims = tuple(self.mnist_train[0][0].shape)
docs/source/extensions/datamodules.rst:155:                # self.dims = tuple(self.mnist_test[0][0].shape)
docs/source/extensions/datamodules.rst:224:                self.dims = self.mnist_train[0][0].shape
docs/source/extensions/datamodules.rst:229:                self.dims = getattr(self, "dims", self.mnist_test[0][0].shape)
pl_examples/basic_examples/mnist_datamodule.py:82:        self.dims = (1, 28, 28)
pl_examples/domain_templates/generative_adversarial_net.py:230:        self.dims = (1, 28, 28)

docs/source/extensions/datamodules.rst:366:        def __init__(self, train_transforms, val_transforms, test_transforms):
docs/source/extensions/datamodules.rst:368:            self.train_transforms = train_transforms
docs/source/extensions/datamodules.rst:369:            self.val_transforms = val_transforms
docs/source/extensions/datamodules.rst:370:            self.test_transforms = test_transforms
docs/source/starter/introduction_guide.rst:1121:    train_transforms = ...
docs/source/starter/introduction_guide.rst:1122:    val_transforms = ...
docs/source/starter/introduction_guide.rst:1123:    test_transforms = ...
pl_examples/basic_examples/mnist_datamodule.py:91:        self.test_transforms = self.default_transforms
pl_examples/basic_examples/mnist_datamodule.py:135:        extra = dict(transform=self.test_transforms) if self.test_transforms else {}
```

File: tests/deprecated_api/test_remove_1-7.py (hunk 4 above)
Comment (ananthsub) [reply tests/deprecated_api/test_remove_1-7.py:69]: ah sorry I missed that below

**Comment by Tshimanga** [comment]
@carmocca so I didn't remove those occurrences because they seemed to be implementation details of those specific datamodules, and I don't believe determining what other datamodules may or may not implement their own additional properties is in scope of this ticket, just cleaning up the `LightningDataModule` interface. 

> train_transforms, val_transforms, test_transforms, dims, and size are entirely optional to use, yet they’re on the DataModule interface. The Trainer does not rely on these, nor is the user forced to implement them. In reality, **these are internal implementation details of individual datamodules**. As a result, we ought to deprecate these off the DataModule interface.

> Of course, **users retain the ability to implement these properties in their LightningModules** if they find these abstractions helpful.

**Comment by Tshimanga** [comment]
By the way, thanks for the attentiveness guys! @carmocca @ananthsub Really appreciate the rapid feedback :)

**Comment by ananthsub** [comment]
> That said, if the goal is actually to purge these properties from any classes implementing the LightningDataModule interface then I can go ahead and do so. Thoughts? @ananthsub

I don't think we need to purge these properties from all classes. If modules want to implement `dims` or `train_transforms` independently, that is their prerogative. as they are implementation details of the module. we just shouldn't expect all modules to have them defined. 

@Tshimanga you could go through the docs to ensure those references to ensure that no arguments are based to the base DataModule constructor.

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by carmocca** [comment]
> @Tshimanga you could go through the docs to ensure those references to ensure that no arguments are based to the base DataModule constructor.

This is important as it's not nice to have the docs using deprecated functionality.

**Comment by Tshimanga** [comment]
@carmocca Good morning, and yeah definitely. I guess I'll open a follow up PR for that after work. Didn't realize this branch would be merged for me. 👍🏽
